Compare USPS to Smarty County FIPs

tamaradement opened the pull request:
## PR Description
Adds an executable script that processes zip-to-county data from CSV, queries the Smarty api, and identifies discrepancies between Smarty's fips for a given 5-digit zip code and the corresponding USPS fips. For now, we output results to the console, and we only log cases where inconsistencies are found.
    
 For context: this pr completes the last step in the following:
- [X] Write a program that leverages the zip4 package to print all counties associated with a zip code. * Addressed in https://github.com/corbaltcode/usps/pull/2
- [X] Output data to CSV format *Addressed in https://github.com/corbaltcode/usps/pull/3
- [X] Use the output or a similar program to find differences between USPS and the [Smarty US ZIP Code API](https://www.smarty.com/docs/cloud/us-zipcode-api). Output data from the usps-to-smarty comparison to CSV or some readable format.

Additional work to be done:
- [ ] Implement a function to convert county codes to county names (optional). This could make data analysis easier.

Example Output:
```
% go run ./cmd/smarty-usps-comparison/main.go -csv "results2.csv"
Zipcode,USPS Fips,Smarty Fips,Mismatch Count,Error
09754,,000,2,Mismatches found: 2
// Other records
```

## PR Checklist
* [n] **New automated tests have been written to the extent possible.**
* [y] **The code has been checked for structural/syntactic validity.**
	- AMI/application: a build was performed
	- terraform changes: "terraform plan" checked on every affected environment
* [n/a] **(If applicable) the code has been manually tested on our infrastructure.**
	- AMI/application: deployed an a test or dev environment
	- terraform changes: applied to test or dev environment
	- script: run against test or dev environment
* [n/a] **Likely failure points and new functionality have been identified and tested manually.**
	Examples:
	- Application manually run in a way that triggers any new branches
	- AMI logged into and changes verified from login shell
* [y] **Pull request description includes a description of all the manual steps performed to accomplish the above.**

To provide feedback on this template, visit https://docs.google.com/document/d/1YfTv7Amyop5G_8w1c2GJ_Mu-70L0KkZHhm9f9umDi3U/edit

Commit:
Adds Smarty Comparison Script

Adds an executable script that processes zip-to-county data from CSV, queries the Smarty api, and identifies discrepancies between Smarty's fips for a given 5-digit zip code and the corresponding USPS fips.

Tested:
Ran the script from start to finish, ensuring all functionalities — from reading from CSV, querying the Smarty api, to processing and comparing fips. Observed the output.

Commit:
Support Standard Input

- Updated csv processing to read from standard input or a specified csv file.
- Added more descriptive directory names.
- Simplified logic in `calculateInconsistencies``
- Reduced the rate limiting delay from 5 seconds to 2 seconds to increase processing speed.
Test Plan:
1. Ran the script with standard input generated from the countylookup  package.
2. Ran the script with a provided csv file path.

jonahb commented on cmd/smarty-usps-comparison/main.go:
```diff
@@ -0,0 +1,163 @@
+	return zips, zipToCounty, nil
+}
+
+func parseCounties(countyData string) []string {
```
This function and parameter could be named more generically since it'll parse any comma-delimited list.

jonahb commented on cmd/smarty-usps-comparison/main.go:
```diff
@@ -0,0 +1,163 @@
+	const rateLimitPause = 2 * time.Second
+
+	for i := 0; i < len(zips); i += batchSize {
+		end := i + batchSize
```
`min` is a builtin function since Go 1.21 ([docs](https://pkg.go.dev/builtin#min)), so another option is:

```
end := min(i + batchSize, len(zips))
```

jonahb commented on cmd/smarty-usps-comparison/main.go:
```diff
@@ -0,0 +1,163 @@
+			end = len(zips)
+		}
+		batch := zips[i:end]
+		responseBody, err := querySmartyAPIBatch(batch, authId, authToken)
```
I might factor the Smarty code differently. A common pattern is a "client" that wraps a service and makes it easy to use in a given language. Here's an [example](https://github.com/corbaltcode/kion/blob/main/internal/client/client.go) from the Kion Tool.

A Smarty client could look something like this:

```
client := smartyzip.NewClient(authID, authToken)
responses, err := client.QueryBatch(zips)
if err != nil { … }

// responses is a []smartyzip.Response
smartyresponseprocessing.ProcessSmartyResponse(responses)
```

(In fact, Smarty offers a [Go SDK](https://www.smarty.com/docs/sdk/go) that includes a client for the US ZIP Code API. I think it's a good exercise to write your own though!)

jonahb commented on internal/smartyresponseprocessing/smartyresponseprocessing.go:
```diff
@@ -0,0 +1,81 @@
+		return fmt.Errorf("failed to unmarshal response: %w", err)
+	}
+
+	for _, response := range smartyResponses {
```
If Smarty thinks a ZIP is invalid, it returns a response of this form ([docs](https://www.smarty.com/docs/cloud/us-zipcode-api)):

```
{
  "input_index": 0,
  "status": "invalid_zipcode",
  "reason": "Invalid ZIP Code."
}
```

We probably want to catch this response and either report it (missing ZIP) or consider it an empty list of counties for the given ZIP.

jonahb commented on internal/smartyresponseprocessing/smartyresponseprocessing.go:
```diff
@@ -0,0 +1,81 @@
+			}
+
+			// Extract the last three digits of the Smarty Fips code, which represent the county.
+			smartyFIPS := []string{zipcode.CountyFIPS[len(zipcode.CountyFIPS)-3:]}
```
Should test `len(zipcode.CountyFIPS)` to avoid a negative index (we can't be sure about that value since it comes from Smarty). Better to fail with an error than panic. Same comment on line 53.

jonahb commented on internal/smartyresponseprocessing/smartyresponseprocessing.go:
```diff
@@ -0,0 +1,81 @@
+	return nil
+}
+
+func calculateInconsistencies(uspsFIPS, smartyFIPS []string) int {
```
This will count the number of counties that are in Smarty but not in the USPS. We want to go the other way too — counties in USPS but not in Smarty. One way to find all diffs would be to sort and compare the slices. The functions in the [slices](https://pkg.go.dev/slices) package (`Sort`, `Equal`, etc.) could help.

jonahb commented on internal/smartyresponseprocessing/smartyresponseprocessing.go:
```diff
@@ -0,0 +1,81 @@
+			inconsistencies := calculateInconsistencies(result.USPSFips, result.SmartyFips)
+			result.Inconsistencies = inconsistencies
+
+			yield(result)
```
This might be more complicated than it needs to be for the volume of data being processed. It would be simpler to return a `[]ZipcodeResult`.

jonahb commented on internal/smartyresponseprocessing/smartyresponseprocessing.go:
```diff
@@ -0,0 +1,81 @@
+package smartyresponseprocessing
```
IMO, it would be okay to fold this in to `smarty-usps-comparison` since it seems specific to that tool. If you're thinking about reusing this logic, though, makes sense as a separate package.

tamaradement replied:
I've opted to fold the `smartyresponseprocessing` package into `smarty-usps-comparison`. I agree that the logic is very tool-specific, so this change makes sense to me. However `main.go` seems a bit lengthy now. In addressing some of the other feedback, I've added a few utility functions which have contributed to its length (some more generic than others.) Would it make sense to create a util folder within `smarty-usps-comparison` to house these?

jonahb replied:
If you think `main.go` is too long, one option is to make other `.go` files in the same package (directory). This is a good option if you don't need the modularity of separate packages (for reuse, information hiding, etc.) but there are logical divisions in the code. For example, the [fmt package](https://github.com/golang/go/tree/master/src/fmt) puts printing, scanning, and formatting each into its own file.

Another option is to split something out into a separate package (directory). In this case, that might make sense for the Smarty stuff. It's easy to imagine using the Smarty client outside this program. A side benefit is that the type names don't need `Smarty`, which is nicer IMO.

tamaradement replied:
Makes sense! If we decide to split some of the smarty-related functionality into a separate package, would belong within the internal directory, or would it be better as a standalone package? I see the potential for it's use outside of `smarty-usps-comparison`  but would it ever be useful outside of the usps module?

jonahb replied:
If we wanted to create a Smarty package for other programs, we'd probably put it in its own module / repo. We'd also have to look harder at the Smarty SDK and decide whether it's worth investing in our own client. So for now, putting this in `internal` makes sense to me.

jonahb reviewed (changes requested):
This looks great. I left a few comments, mostly about code organization. Glad to talk it over if that would help.

I haven't run the tool, but I agree it'd be good to identify or exclude military addresses. The `zipcode_type` field of the response could help.

Commit:
Updates parseCounties to parseCSVList.

Since parseCounties will parse any comma-delimited list, it makes sense to have a more generic function name and variable names.

Tested by running the script with -csv flag.

Commit:
Update Slice Comparison to Use Boolean Indicator

Refactored the calculateInconsistencies function to slicesDiffer.
This update simplifies the approach by using a boolean to indicate the
presence of mismatches between the smarty and usps slices,
rather than counting each discrepancy.

Tested: Executed the updated script to verify
that the new boolean approach correctly identifies when mismatches are present.

Commit:
Implemented Smarty Client

Adds implementation for a `SmartyClient` in order to encapsulate all Smarty-related query logic and response handling.

Tested: Executed the modified script to ensure the new client properly handles api requests.

Commit:
Integrates `smartyresponseprocessing` Into main

This change folds `smartyresponseprocessing` package directly into the `smarty-usps-comparison/main.go` file in order to consolidate all tool-specific logic into one main script.

Tested: Executed the script to confirm that the integrated processing logic functions correctly.

Commit:
Refactor ProcessSmartyResponse for better
error handling and output clarity.

- Changed `hasMismatch` to `ErrorMessage` to include detailed error messages indicating either mismatch count or invalid input.
- Added safety checks to ensure that fips codes are at least three characters long before extracting the last three digits.
- Improved the output formatting .

Tested the modifications by running the script and checking that the output is correctly formatted and errors/mismatches are accurately reported.

Commit:
Simplified iterative batch logic.

Use Go's built in min function to simply iterating over batches of query inputs.

Tested the modifications by running the script and checking that the output is correctly formatted and mismatches are reported.

Commit:
Adds Safety Checks to ProcessSmartyResponse

- Adds handling for an invalid_zipcode response.
- Adds a check to verify that response.InputIndex is a valid index.
- Adds a check to verify that county FIPS strings have at least 3 characters.

Tested:Tested the modifications by running the script and checking that the output is correctly formatted and mismatches are reported.

Commit:
Adds `smarty` internal package

- Houses logic specific to smarty api interactions into a dedicated internal package.
- Updates go version in go.mod from 1.17 to 1.21 in order utilize go's built-in min function.

Tested:
- Ran each command in cmd to ensure they were still working as expected with Go's version update.
- Ran the smarty-usps-comparison tool to ensure discrepancies between USPS and Smarty data were still being appropriately logged.

Commit:
Updates wording of error logging in smarty

Updates wording of error logging in two places within the smarty package.

jonahb commented on cmd/smarty-usps-comparison/main.go:
```diff
@@ -0,0 +1,229 @@
+	return ""
+}
+
+func countMismatches(uspsFIPS, smartyFIPS []string) int {
```
This is a nice algorithm! Similar to `parseCSVList`, the function would work on any two string slices, so you could make the parameter names more generic.

jonahb commented on cmd/smarty-usps-comparison/main.go:
```diff
@@ -0,0 +1,229 @@
+}
+
+func main() {
+	authId := mustGetenv("AUTH_ID")
```
It would be good to move lines 163-165 after `flag.Parse()` so that you get usage if you run the program with `--help`, even if the env vars are unset.

jonahb commented on cmd/smarty-usps-comparison/main.go:
```diff
@@ -0,0 +1,229 @@
+	"github.com/corbaltcode/usps/internal/smarty"
+)
+
+type ZipcodeResult struct {
```
Would it be worth preserving the number of mismatches as an integer, e.g. `MismatchCount int`, and including this in the output?

tamaradement replied:
I wasn't sure how valuable it was to keep the mismatchCount field because it's now baked into the errorMessage field. But now, I think I see advantages to doing this. It will be more straightforward when analyzing the output programmatically if we don't have to parse the error message string.

jonahb replied:
Yep, and I think it'll also help if you import the output into a spreadsheet. Then you can sort by that column.

jonahb commented on cmd/smarty-usps-comparison/main.go:
```diff
@@ -0,0 +1,229 @@
+	return zips, zipToCounty, nil
+}
+
+func ProcessSmartyResponse(responseBody []smarty.SmartyResponse, zipToCounties map[string][]string, zipcodes []string, startIndex int) ([]ZipcodeResult, error) {
```
This interface is hard to understand because some of the names are generic, it has a lot of parameters, and it's not clear how the parameters are used (e.g. `startIndex`). You have to read the implementation to know what it does.

One way to clarify things would be to avoid generic words like "process" and "result" (generally a good idea). As an example, `ZIPCountyDiff` might be more descriptive than `ZipcodeResult`.

But also, you may want to make this interface "narrower" (fewer parameters). For example, if the function compared a single Smarty response, you could do in `main` (pseudocode):

```
// query smarty
// verify len(smartyResponses) == len(batch)

for j, resp := range smartyResponses {
  zip := batch[j]
  diff := diffSmarty(zip, zipToCounty[zip], resp)
  // write diff to csv
}
```

I'm not sure, but I think this would reduce the amount of code and make it clearer and less brittle as well.

tamaradement replied:
Thanks for this feedback! I agree with the suggestions, especially about simplifying the function interface. The pseudocode also makes sense to me. Regarding the check on `len(smartyResponses) == len(batch)`:  in the case of the last batch, we might have fewer responses than the given `batchSize`. So I updated this part of the code to check that the length of responses doesn't exceed `batchSize`, so we don't have out-of-bounds type errors.

jonahb replied:
That works to avoid out-of-bounds errors. I think even in the last batch, `len(batch) == len(responseBody)`. (`len(batch)` may be less than `batchSize` due to lines 194-195.)

tamaradement replied:
Oh, right! Will change to `==`

jonahb commented on internal/smarty/smarty.go:
```diff
@@ -0,0 +1,101 @@
+	Zipcode string `json:"zipcode"`
+}
+
+type SmartyResponse struct {
```
You may want to omit `Smarty` from the public members of the package. See "Naming package contents" in this Go blog post: https://go.dev/blog/package-names#naming-package-contents.

jonahb commented on internal/smarty/smarty.go:
```diff
@@ -0,0 +1,101 @@
+	State             string `json:"state"`
+}
+
+type SmartyClient struct {
```
Do users of this type need access to the members after initialization? If not, good to make them private.

jonahb replied:
Ah, I meant that you may want to make the fields private. I was thinking especially of `BaseURL`, which you probably don't want a client mucking with.

It also works to make `client` private and return it from a public function (`NewClient`), though I'm not sure I've seen this pattern much.

jonahb commented on internal/smarty/smarty.go:
```diff
@@ -0,0 +1,101 @@
+		return nil, fmt.Errorf("failed to marshal JSON payload: %w", err)
+	}
+
+	apiURL := fmt.Sprintf("%s?auth-id=%s&auth-token=%s", client.BaseURL, client.AuthId, client.AuthToken)
```
It's a good idea to escape values inserted into a query string in case they include reserved characters. One way to do this is `QueryEscape` in `net/url`. I also like `url.Values`, e.g.:

```
v := url.Values{}
v.Add("auth-id", client.AuthId)
…
url := client.baseURL + "?" + v.Encode()
```

jonahb reviewed (changes requested):
This looks great. I left a few more comments. I think two of them should be addressed before you merge — the query escaping and allowing `--help` without env vars set. The others are about readability / maintainability, so feel free to defer (or push back!). Also glad to discuss if the comments don't make sense.

Commit:
Adds Mismatch Count to Output

type ZipcodeResult struct {
	Zipcode       string
	USPSFips      []string
	SmartyFips    []string
	MismatchCount int
	ErrorMessage  string
}

- Tested by running the script and checking the results to see that all fields are appropriately logged.

Commit:
Moves Flag Parsing Closer to Top

- Moves flag parsing closer to top of main to enable users to run program with --help,
even if env variables are not yet set.
- Adds more generic parameter names for `countMismatches`

Tested: Ran the program with --help.

Commit:
Refactor for ProcessSmartyResponse

- Narrowed the interface of ProcessSmartyResponse to take fewer parameters.
- Renamed the function for clarity.
- Broke up the functionaility into two utilities:
  - `generateZipcodeCountyDiff`
  - `extractFipsCodes`

Tested by running the script and checking the results to see that all fields are appropriately logged.

Commit:
Omit Smarty from Public Members

- Removes smarty from the names of public members from the smarty package.

Tested by running the script and checking the results to see that all fields are appropriately logged.

Commit:
Make Certain Smarty Members Private

- Sets smarty members not required in other packages to private.

Tested by running the script and checking the results to see that all fields are appropriately logged.

Commit:
Adds encoding for apiURL

Refactor URL query parameter handling to use url.Values for automatic encoding.

Tested by running the script and checking the results to see that all fields are appropriately logged.

Commit:
Updated variable names in `extractCountyFipsCodes`

- Changed extractFipsCodes to extractCountyFipsCodes
- Updated parameter name and internal variable for enhanced clarity

Tested by running the script and checking the results to see that all fields are appropriately logged.

jonahb reviewed (approved):
Looks great! Much easier to read than the last revision. It'll be interesting to run this on the latest Smarty.

Commit:
Update response count check and made client fields private

- Modified the condition in the main function from checking if the number of responses exceeds the number of requests to checking if the number of responses does not exactly match the number of requests.
- Made all fields in the Client struct private.

Testing:
Successfully ran the script from start to finish, verifying that all batches, including the last batch, were correctly processed and recorded without any issues.
